server/api/rounds.py

```diff
@@ -1033,3 +1033,37 @@ def list_rounds_jurisdiction_admin(
     election: Election, jurisdiction: Jurisdiction  # pylint: disable=unused-argument
 ):
     return jsonify({"rounds": [serialize_round(r) for r in election.rounds]})
+
+
+@api.route("/election/<election_id>/round/<round_id>", methods=["DELETE"])
+@restrict_access([UserType.AUDIT_ADMIN])
+def undo_create_round(election: Election, round: Round):
+    current_round = get_current_round(election)
+    if not current_round or current_round.id != round.id:
+        raise Conflict(
+            "Cannot undo starting this round because it is not the current round."
+        )
+
+    if len(list(round.audit_boards)) > 0:
+        raise Conflict(
+            "Cannot undo starting this round because some jurisdictions have already created audit boards."
+        )
+
+    db_session.delete(round)
+    # Delete any sampled ballots that were created this round (they will have no
+    # associated SampledBallotDraws, since they are deleted by cascade when
+    # deleting the round).
+    SampledBallot.query.filter(
+        SampledBallot.id.in_(
+            SampledBallot.query.join(Batch)
+            .join(Jurisdiction)
+            .filter_by(election_id=election.id)
+            .filter(not_(SampledBallot.draws.any()))
+            .with_entities(SampledBallot.id)
+            .subquery()
+        )
+    ).with_entities(SampledBallot).delete(synchronize_session=False)
+
+    db_session.commit()
```

Comment:
I *think* there is a potential race condition - albeit an unlikely one - if a jurisdiction creates audits boards right after the check for audit boards but before the DB call to delete the round.

The only way to make this strictly work is to ensure both are locking on the same entity and do a `select for update` here on that/those entity/ies.

Thoughts?

Reply:
I agree there's a race condition, and that it's fairly unlikely. I believe if this occurred, the audit boards that had just gotten created would instantly be deleted by cascade, since they reference the round. I think that's ok behavior in this case, since the jurisdiction wouldn't have been able to actually start auditing in that very short moment. It might even be better behavior because then the audit admin that was trying to undo launch can continue doing that, instead of getting blocked.

Reply:
👍